### PR TITLE
Add streaming match generation

### PR DESCRIPTION
**Pull request by davidmezzetti**
This PR proposes "streaming match generation" for `grandiso`.

When working with large graphs and broad queries, the project currently has a number of places where results are queued up in lists. This requires having the full list at each step before proceeding to the next step. On top of that, if a `LIMIT` clause is specified, in some cases a large list needs to be built before the LIMIT clause is recognized and execution can be terminated.

This PR is as minimally invasive to the current code as possible. In most cases, the iteration logic is replaced with an inner function that runs a generator. When these generators are chained together, it effectively walks through the graph iteratively and enables returning results in a "streaming" fashion. 

There will be a corresponding PR in the GrandCypher project that will have a couple examples illustrating the full change.

### CODE REVIEW

File: grandiso/__init__.py [hunk 1]
```diff
@@ -298,49 +294,49 @@ def get_next_backbone_candidates(
...
-            # the equivalent edge in the host graph:
-            if not motif.has_edge(motif_u, motif_v) and host.has_edge(
-                result[motif_u], result[motif_v]
-            ):
-                # this is a violation.
-                break
-        else:
-            isomorphism_candidates.append(result)
-    return isomorphism_candidates
+    def isomorphism_candidates():
```
Comment (j6k4m8) [inline grandiso/__init__.py:324]: sorry for the quiet; doing a little benchmarking right now on the performance implications of calling a nested func vs unwrapping these!

Comment (davidmezzetti) [reply grandiso/__init__.py:324]: Sounds good. I tried to make the change as minimally invasive as possible but I agree that having them as separate functions would be cleaner.

Let me know if you come across anything that seems slower. In the worst case scenario, it goes through all the data and discovers there are no matches (i.e. a full scan). 

I also think the grand project is a pretty cool idea, putting a networkx interface on different backends such as a SQL database. I could see where indexes could make full scan operations much faster.

Comment (j6k4m8) [reply grandiso/__init__.py:324]: @davidmezzetti I got about +10μs benchmarking locally with ASV, which I think is fair to say is in the noise :) nice work!

I'd LOVE to have better indexing in Grand; I've been aiming for feature-completeness first and then adding in "shortcuts" and optimizations; would love to chat about some gameplans if you're interested!

Comment (davidmezzetti) [reply grandiso/__init__.py:324]: Sounds good to me. I think having a networkx dialect applied towards something like a SQLite backend with indexes could be a powerful in-memory graph solution.

Things like an attribute search could return immediately regardless if there was/wasn't data for example (i.e. no full scans). In the case of SQLite, you'd get the native code and smart memory management for free. But still have the ease-of-use of networkx.

Comment (j6k4m8) [reply grandiso/__init__.py:324]: COMPLETELY agree, would love to have a sqlite true high speed implementation!

**Pull request merged**